pkg/api/customization/pipeline/cluster_pipeline.go

```diff
@@ -0,0 +1,326 @@
+	clusterPipeline, err := clusterPipelines.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if authAppInput.SourceCodeType == "github" {
+		clusterPipeline.Spec.GithubConfig = &v3.GithubClusterConfig{
+			TLS:          authAppInput.TLS,
+			Host:         authAppInput.Host,
+			ClientID:     authAppInput.ClientID,
+			ClientSecret: authAppInput.ClientSecret,
+			RedirectURL:  authAppInput.RedirectURL,
+		}
+	}
+	//oauth and add user
+	userName := apiContext.Request.Header.Get("Impersonate-User")
+	if _, err := h.authAddAccount(clusterPipeline, authAppInput.SourceCodeType, userName, authAppInput.RedirectURL, authAppInput.Code); err != nil {
+		return err
+	}
+	//update cluster pipeline config
+	if _, err := h.Management.Management.ClusterPipelines(ns).Update(clusterPipeline); err != nil {
+		return err
+	}
+
+	data := map[string]interface{}{}
+	if err := access.ByID(apiContext, apiContext.Version, apiContext.Type, apiContext.ID, &data); err != nil {
+		return err
+	}
+
+	apiContext.WriteResponse(http.StatusOK, data)
+	return nil
+}
+
+func (h *ClusterPipelineHandler) authuser(apiContext *types.APIContext) error {
+
+	parts := strings.Split(apiContext.ID, ":")
+	if len(parts) <= 1 {
+		return errors.New("invalid ID")
+	}
+	ns := parts[0]
+	name := parts[1]
+
+	authUserInput := v3.AuthUserInput{}
+	requestBytes, err := ioutil.ReadAll(apiContext.Request.Body)
+	if err != nil {
+		return err
+	}
+	if err := json.Unmarshal(requestBytes, &authUserInput); err != nil {
+		return err
+	}
+
+	clusterPipelines := h.Management.Management.ClusterPipelines(ns)
+	clusterPipeline, err := clusterPipelines.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if authUserInput.SourceCodeType == "github" && clusterPipeline.Spec.GithubConfig == nil {
+		return errors.New("github oauth app is not configured")
+	}
+
+	//oauth and add user
+	userName := apiContext.Request.Header.Get("Impersonate-User")
+	logrus.Debugf("try auth with %v,%v,%v,%v,%v", clusterPipeline, authUserInput.SourceCodeType, userName, authUserInput.RedirectURL, authUserInput.Code)
+	account, err := h.authAddAccount(clusterPipeline, authUserInput.SourceCodeType, userName, authUserInput.RedirectURL, authUserInput.Code)
+	if err != nil {
+		return err
+	}
+	data := map[string]interface{}{}
+	if err := access.ByID(apiContext, apiContext.Version, client.SourceCodeCredentialType, account.Name, &data); err != nil {
+		return err
+	}
+	apiContext.WriteResponse(http.StatusOK, data)
+	return nil
+}
+
+func (h *ClusterPipelineHandler) revokeapp(apiContext *types.APIContext) error {
+
+	parts := strings.Split(apiContext.ID, ":")
+	if len(parts) <= 1 {
+		return errors.New("invalid ID")
+	}
+	ns := parts[0]
+	name := parts[1]
+
+	clusterPipelines := h.Management.Management.ClusterPipelines(ns)
+	clusterPipeline, err := clusterPipelines.Get(name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	if err := h.cleanup(ns, apiContext); err != nil {
```

Comment:
We should try to avoid putting too much login into api package; putting it to controller is a preferable way. API layer should just set the field, and invoke the update for the underlying resource. Then the controller for the resource should listen and react on update events.

Reply:
moved to controllers